java/src/main/java/org/apache/lucene/analysis/hebrew/TokenFilters/HebrewLemmatizerTokenFilter.java

```diff
@@ -0,0 +1,150 @@
+        super(input);
+        this.dict = dict;
+        this.lemmatizer = new Lemmatizer(dict);
+        this.lemmatizeExactHebrewWords = lemmatizeExactHebrewWords;
+        this.lemmatizeExactNonHebrewWords = lemmatizeExactNonHebrewWords;
+    }
+
+    @Override
+    public boolean incrementToken() throws IOException {
+        // first check if we have leftover lemmas
+
+        if (!previousLemmas.isEmpty()) {
+            clearAttributes();
+            String tokenVal;
+            // Index all unique lemmas at the same position
+            if (previousType == HebrewTokenTypeAttribute.HebrewType.Hebrew ||
+                    previousType == HebrewTokenTypeAttribute.HebrewType.Acronym ||
+                    previousType == HebrewTokenTypeAttribute.HebrewType.Construct) {
+                HebrewToken hebToken = (HebrewToken) previousLemmas.remove(0);
+                tokenVal = hebToken.getLemma() == null ? hebToken.getText().substring(hebToken.getPrefixLength()) : hebToken.getLemma();
+            } else {
+                tokenVal = previousLemmas.remove(0).getText();
+            }
+            termAtt.setEmpty().append(tokenVal);
+            hebrewTypeAtt.setType(HebrewTokenTypeAttribute.HebrewType.Lemma);
+            posIncrAtt.setPositionIncrement(0);
+            offsetAtt.setOffset(previousStartOffset, previousEndOffset);
+            return true;
+        }
+
+        // if input is over, so are we
+        if (!input.incrementToken()) {
+            return false;
+        }
+        // we do not lemmatize numbers or exact words (if specified)
+        if (hebrewTypeAtt.isNumeric() || (hebrewTypeAtt.isExact() &&
+                ((!lemmatizeExactHebrewWords && hebrewTypeAtt.isHebrew()) ||
+                        (!lemmatizeExactNonHebrewWords && hebrewTypeAtt.getType () == HebrewTokenTypeAttribute.HebrewType.NonHebrew))))
+        {
+            return true;
+        }
+
+        previousLemmas.clear();
+        duplicateLemmas.clear();
+        previousStartOffset = offsetAtt.startOffset();
+        previousEndOffset = offsetAtt.endOffset();
+        previousType = hebrewTypeAtt.getType();
+
+        // Handle hebrew words
+        if (hebrewTypeAtt.isHebrew()) {
+            previousTolerated = false;
+
+            String word = termAtt.toString();
+
+            // try to lemmatize
+            List<HebrewToken> tempList = lemmatizer.lemmatize(word);
+            // word wasn't found in the dictionary - try tolerating it
+            if (tempList.isEmpty()) {
+                tempList = lemmatizer.lemmatizeTolerant(word);
+                previousTolerated = true;
+            }
+            // add words to the previousLemmas : remove duplicates and tokens which aren't ranked high enough if they came from tolerating.
+            // TODO: consider the ranking as an additional filter
```

Comment:
What was? Ranking? This is done at the Lemmatizer. The logic that ranks is done while lemmatizing, so making it into a another filter probably isn't the way to go.